Firestore/core/src/firebase/firestore/core/view_snapshot.h

```diff
@@ -89,8 +100,128 @@ class DocumentViewChangeSet {
+   * added.
+   */
+  static ViewSnapshot FromInitialDocuments(FSTQuery* query,
+                                           FSTDocumentSet* documents,
+                                           model::DocumentKeySet mutated_keys,
+                                           bool from_cache,
+                                           bool excludes_metadata_changes);
+
+  /** The query this view is tracking the results for. */
+  FSTQuery* query() const {
+    return impl_->query;
+  }
+
+  /** The documents currently known to be results of the query. */
+  FSTDocumentSet* documents() const {
+    return impl_->documents;
+  }
+
+  /** The documents of the last snapshot. */
+  FSTDocumentSet* old_documents() const {
+    return impl_->old_documents;
+  }
+
+  /** The set of changes that have been applied to the documents. */
+  const std::vector<DocumentViewChange>& document_changes() const {
+    return impl_->document_changes;
+  }
+
+  /** Whether any document in the snapshot was served from the local cache. */
+  bool from_cache() const {
+    return impl_->from_cache;
+  }
+
+  /** Whether any document in the snapshot has pending local writes. */
+  bool has_pending_writes() const {
+    return !(impl_->mutated_keys.empty());
+  }
+
+  /** Whether the sync state changed as part of this snapshot. */
+  bool sync_state_changed() const {
+    return impl_->sync_state_changed;
+  }
+
+  /** Whether this snapshot has been filtered to not include metadata changes */
+  bool excludes_metadata_changes() const {
+    return impl_->excludes_metadata_changes;
+  }
+
+  /** The document in this snapshot that have unconfirmed writes. */
+  model::DocumentKeySet mutated_keys() const {
+    return impl_->mutated_keys;
+  }
+
+  std::string ToString() const;
+  size_t Hash() const;
+
+  friend bool operator==(const ViewSnapshot& lhs, const ViewSnapshot& rhs);
+
+ private:
+  struct Impl {
+    Impl(FSTQuery* query,
+         FSTDocumentSet* documents,
+         FSTDocumentSet* old_documents,
+         std::vector<DocumentViewChange> document_changes,
+         bool from_cache,
+         bool sync_state_changed,
+         bool excludes_metadata_changes,
+         model::DocumentKeySet mutated_keys)
+        : query{query},
+          documents{documents},
+          old_documents{old_documents},
+          document_changes{std::move(document_changes)},
+          from_cache{from_cache},
+          sync_state_changed{sync_state_changed},
+          excludes_metadata_changes{excludes_metadata_changes},
+          mutated_keys{std::move(mutated_keys)} {
+    }
+
+    FSTQuery* query = nil;
+
+    FSTDocumentSet* documents = nil;
+    FSTDocumentSet* old_documents = nil;
+    std::vector<DocumentViewChange> document_changes;
+
+    bool from_cache = false;
+    bool sync_state_changed = false;
+    bool excludes_metadata_changes = false;
+
+    model::DocumentKeySet mutated_keys;
+  };
+
+  std::shared_ptr<const Impl> impl_;
```

Comment:
Note: this is purely an optimization. There are situations where copying a snapshot is unavoidable, and it could be a large object, especially once `documents` and `old_documents` are C++ as well.

Reply:
@rsgowman We could in theory implement something like small buffer optimization, but it seems like a huge overkill at this point.

@wilhuff Gil, I'd appreciate your input on `shared_ptr` here as well.

Reply:
As long as you're asking my opinion... this entire conversation is venturing into overkill territory: we're a database! We do disk and network I/O. Shaving a few cycles here and there almost doesn't matter. Readability, simple code, and similarity to the other ports should always be our top priorities.

With that said, here's some specific feedback to the points raised here:

**Large object**: we have already made copies of building block types cheap. For example: `model::DocumentKeySet` is an `immutable::SortedSet<DocumentKey>`. `FSTDocumentSet` is currently backed by objective-c immutable sorted set and maps and the port will be backed by C++ equivalents. We shouldn't worry about whether or not types composed of these types copy them.

**`std::shared_ptr` is slow** isn't that useful an argument. Every pointer in the Objective-C implementation before we started was equivalent either to a `std::shared_ptr` or a `std::weak_ptr`. We have little reason to go out of our way to improve upon this because it was already fast enough. When building the `immutable::LlrbNode` stuff I benchmarked single threaded construction and destruction of thousands of nodes and `shared_ptr` was only 30% slower than `unique_ptr`. While it is undoubtedly slower in an absolute sense, for our application it doesn't matter.

A good argument against `std::shared_ptr` is that makes the ownership or destruction order confusing or that a simpler pointer type would suffice, especially with a clear ownership model that makes those kinds of pointers work. In cases where data's immutable and ownership is unclear, `shared_ptr` is totally reasonable and we shouldn't think about the cost.

**small buffer optimization** I don't think we should implement things like this until a profiler shows us it's critical to do so. The code complexity of doing so is massive, and we should only do so for critical internal types, rather than one like this which is essentially just a named, typed tuple.

In the larger context, `ViewSnapshot` objects are transient collections of data composed from possibly shared data. They exist to bundle everything up required to generate a snapshot, but they don't live very long. I'm not willing to go so far as to make this just a dumb struct, but making this just a bundle of private values and const getters seems sufficient.

Reply:
Gil, that's a good point about `FSTDocumentSet`. A vector of `DocumentViewChange`s could potentially be large, but it becomes less convincing at this point. I've reverted the `shared_ptr`.